orderer/common/cluster/block_fetcher.go

```diff
@@ -0,0 +1,522 @@
+		// don't append currentEndpoint and suspects
+		if bf.suspects.has(e.Endpoint) || e.Endpoint == bf.currentEndpoint.Endpoint {
+			continue
+		}
+		candidates = append(candidates, e)
+	}
+	// select a random entry from candidates
+	bf.currentEndpoint = groupCollection(candidates).select_random_entries(1)[0]
+
+	bf.Logger.Debugf("Shuffled block puller endpoint. Current endpoint: [%s]", bf.currentEndpoint)
+	// Update current blocksource
+	bf.setBlockSource(bf.currentEndpoint)
+	bf.LastShuffledAt = bf.TimeNow()
+	return nil
+}
+
+func (bf *BlockFetcher) isShuffleTime() bool {
+	if bf.ShuffleTimeout == time.Duration(0) {
+		// It should return only false if Time based Shuffling is not enabled
+		bf.Logger.Debugf("Shuffle Disabled.")
+		return false
+	}
+	shuffle := bf.TimeNow().After(bf.LastShuffledAt.Add(bf.ShuffleTimeout))
+	bf.Logger.Debugf("Shuffle Enabled. Time to shuffle %v", shuffle)
+	return shuffle
+}
+
+//isEndpointByzantine connects to BlockFetcher.MaxByzantineNodes orderer nodes and pulls attestation blocks from them
+//to suspect the byzantine behaiour of orderer nodes.
+func (bf *BlockFetcher) isEndpointByzantine(ec EndpointCriteria, seq uint64) (bool, error) {
+	bf.Logger.Infof("Checking for BFT Behavior of [%s]", ec.Endpoint)
+
+	candidates := []EndpointCriteria{}
+	for _, e := range bf.Config.Endpoints {
+		// don't append currentEndpoint and suspects
+		if bf.suspects.has(e.Endpoint) || e.Endpoint == bf.currentEndpoint.Endpoint {
+			continue
+		}
+		candidates = append(candidates, e)
+	}
+
+	group := groupCollection(candidates)
+
+	wg := sync.WaitGroup{}
+
+	// attestationsC is a channel to collect attestation blocks
+	attestationsC := make(chan *orderer.BlockAttestation, bf.MaxByzantineNodes)
+
+	for i := uint64(0); i < uint64(bf.MaxByzantineNodes); i++ {
+		wg.Add(1)
+
+		// goroutine to pull attestations and put then in attestationsC channel
+		go func(c chan *orderer.BlockAttestation) {
+			defer wg.Done()
+			config := bf.Config
+			var attestation_puller AttestationSource
+
+			// This `for` loop should not go infinite, so looping for maximum
+			// (uint64(len(group)) + uint64(1) - bf.MaxByzantineNodes) times
+			// e.g: For Nodes=10, MaxByzantineNodes=3, 1st groutine selects an orderer and pulls attestations,
+			// then 2nd goroutine successfully pulls attestations, then we have 10 - (3 - 1) orderers to select from
+			// for the 3rd goroutine, so assuming pulling from orderers fails, we can try for maximum 10 - (3 - 1) times,
+			// unless we find a unique orderer which sends attestations.
+			for j := uint64(0); j < (uint64(len(group)) + uint64(1) - bf.MaxByzantineNodes); j++ {
+				// Choose a random orderer node and and try to pull attestations from it.
+				// if not able to pull attestation s due to error, choose a different
+				// endpoint, until an endpoint is found which sends attestations.
+
+				// choose a random orderer node
+				config.Endpoints = group.select_random_entries(1)
```

Comment:
what if you choose the same node in different goroutines?

Reply:
the random function will return unique endpoints, the random function will just be iterating over the collection to make sure that values are unique